src/vector_calculus.f90

```diff
@@ -0,0 +1,346 @@
+module m_vector_calculus
+  use m_allocator, only: allocator_t, field_t
+  use m_base_backend, only: base_backend_t
+  use m_common, only: dp, DIR_X, DIR_Y, DIR_Z, &
+                      RDR_X2Y, RDR_X2Z, RDR_Y2X, RDR_Y2Z, RDR_Z2X, RDR_Z2Y
+  use m_tdsops, only: tdsops_t
+
+  implicit none
+
+  type :: vector_calculus_t
+    !! Defines vector calculus operators
+    class(base_backend_t), pointer :: backend
+  contains
+    procedure :: curl
+    procedure :: divergence_v2p
+    procedure :: gradient_p2v
+    procedure :: laplacian
+  end type vector_calculus_t
+
+  interface vector_calculus_t
+    module procedure init
+  end interface vector_calculus_t
+
+contains
+
+  function init(backend) result(vector_calculus)
+    implicit none
+
+    class(base_backend_t), target, intent(inout) :: backend
+    type(vector_calculus_t) :: vector_calculus
+
+    vector_calculus%backend => backend
+
+  end function init
+
+  subroutine curl(self, o_i_hat, o_j_hat, o_k_hat, u, v, w, &
+                  x_der1st, y_der1st, z_der1st)
+    !! Curl of a vector field (u, v, w).
+    !! Inputs from velocity grid and outputs to velocity grid.
```

Comment:
Shall we use the `CELL`/`FACE` etc. nomenclature instead of velocity/pressure grids? Just to be consistent and general